### ISSUE TITLE
[`shopify-app-express`]: validateAuthenticatedSession for non-embedded apps that make XHR requests for data?

### ISSUE THREAD

**Issue by justinhelmer** [Shopify/shopify-app-js#470]
## Overview

👋 Thanks for supporting this robust SDK that does a lot of the heavy lifting for us!

We are building a straightforward application that leverages Shopify APIs on the backend.

After the initial installation and scope grant, we want to make subsequent calls to the API that are instantiated by a user interaction in the frontend. For this purpose, we want to use a short-lived (online) access token as described by the [docs](https://shopify.dev/docs/apps/auth/oauth/access-modes#online-access):

> Online access is meant to be used when a user is interacting with your app through the web, [...]
>
> Example use case(s): Your app's security requirements specify short-lived access to a store. Tokens with online access mode expire either when the user logs out or after 24 hours.

To illustrate the use-case:

1. User installs Shopify app, accepts grant, and is taken to our product landing page (self hosted; not embedded).
2.  On the product landing page, the user can subscribe to a _Billing Plan_ of their choosing.
3. The "Subscribe" action (triggered by the user) makes a call to our backend (e.g. `/api/pay?store={name}`).
4. The backend endpoint validates that the session is valid:
    4a. If valid, the route handlers makes the M2M call to Shopify's billing API, and respond accordingly
    4b. If not valid, the route handler results in a redirection to begin the authn/authz flow. Once authorized, they will then be redirected back to our product page, which should then automatically trigger the `/api/pay?store={name}` request again (this time with valid credentials).

For (4), on the surface, it would appear that this is specifically what the [validateAuthenticatedSession](https://github.com/Shopify/shopify-app-js/blob/bdfd30bea4b743d61e8b3589a48021b6c0411399/packages/shopify-app-express/src/middlewares/validate-authenticated-session.ts#L13) is designed to do. The [documentation](https://github.com/Shopify/shopify-app-js/blob/bdfd30bea4b743d61e8b3589a48021b6c0411399/packages/shopify-app-express/docs/reference/validateAuthenticatedSession.md) suggests it should behave correctly for both embedded and non-embedded apps.

In the case where the user is authorized, the middleware works great. However, in the case where the session is invalid, the [redirectOutOfApp](https://github.com/Shopify/shopify-app-js/blob/main/packages/shopify-app-express/src/redirect-out-of-app.ts#L9) function will result in a `302` to the auth endpoint. This tells the browser (who initialized the `fetch()` request) to redirect. The first redirect to the auth endpoint is fine, but then _that_ route handler redirects to `/admin/oauth/authorize` on the shopify store domain. Since this is a cross-origin redirect, the browser fails due to CORS:

```
Request URL:
http://localhost:8445/api/integration/shopify/pay?shop={shop-name}.myshopify.com&plan=Plus%20Annually
Request Method:
GET
Status Code:
302 Found
--> Location: /api/integration/shopify/auth-v2?shop={shop-name}.myshopify.com

Request URL:
http://localhost:8445/api/integration/shopify/auth-v2?shop={shop-name}.myshopify.com
Request Method:
GET
Status Code:
302 Found
--> Location: https://{shop-name}.myshopify.com/admin/oauth/authorize?client_id={client-id}&scope=write_themes%2Cread_products%2Cwrite_pixels%2Cread_customer_events&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fapi%2Fintegration%2Fshopify%2Fauth-v2%2Fcallback&state=147754441580872&grant_options%5B%5D=

Request URL:
https://{shop-name}.myshopify.com/admin/oauth/authorize?client_id={client-id}&scope=write_themes%2Cread_products%2Cwrite_pixels%2Cread_customer_events&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fapi%2Fintegration%2Fshopify%2Fauth-v2%2Fcallback&state=147754441580872&grant_options%5B%5D=
--> ERROR
```
> Access to fetch at 'https://{shop-name}.myshopify.com/admin/oauth/authorize?client_id={client-id}&scope=write_themes%2Cread_products%2Cwrite_pixels%2Cread_customer_events&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fapi%2Fintegration%2Fshopify%2Fauth-v2%2Fcallback&state=147754441580872&grant_options%5B%5D=' (redirected from 'http://localhost:8445/api/integration/shopify/pay?shop={shop-name}.myshopify.com&plan=Plus%20Annually') from origin 'http://localhost:8445' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

-------------------------

To work around this, I have a custom middleware that is essentially a copy/paste of [validateAuthenticatedSession](https://github.com/Shopify/shopify-app-js/blob/bdfd30bea4b743d61e8b3589a48021b6c0411399/packages/shopify-app-express/src/middlewares/validate-authenticated-session.ts#L13). The only difference is that it responds with an HTTP `401` or `403` (with the `redirectUri` in the response body) as opposed to a `302`. This gives the XHR call a chance to react and set the `window.location` to the `redirectUri`, because there’s no mechanism for `fetch()` to capture a 3xx redirect response and handle it manually (as it would violate [atomic HTTP redirect handling](https://fetch.spec.whatwg.org/#atomic-http-redirect-handling) in the case where the URI exposes sensitive data).

-------------------------

For the case of non-embedded apps that make XHR calls, it would be nice to have an out-of-the-box middleware, especially since the current middleware is 90% there.

Besides the need to be able to change the `window.location`, we would also need some special redirect handling (e.g. using `/my-home-page` vs `/my-home-page?just-authorized` as the `callbackPath`), so that once the auth completes, the product page would then know to trigger the request initiated by step (3) "Subscribe" above.

-------------------------

I would also have expected this to be a fairly straightforward / common case (i.e. making credentialed requests via XHR in a non-embedded app), so I am surprised there is not something more well established. It would seem the App Bridge capability is tied to embedded apps w/ session tokens. This leads me to believe I may be missing something simple.

Thoughts?

**Comment by justinhelmer** [comment]
For the sake of completeness, below is the server-side `/pay` (express) route handler described in the original post. You will find it to be nearly _identical_ to the built-in [validateAuthenticatedSession](https://github.com/Shopify/shopify-app-js/blob/bdfd30bea4b743d61e8b3589a48021b6c0411399/packages/shopify-app-express/src/middlewares/validate-authenticated-session.ts#L13) middleware (for the non-embed case), with the exception that we respond with a `403` (and `redirectUrl`) as opposed to a `302`. 

The other addition is the middleware that sets the "state" cookie (in this case, `plan`), so that the client knows to trigger the request again after the authn/authz flow completes:

```typescript
router.get(
  '/pay',

  // Input validation of query parameters
  filterQuery({ shop: iv.STRING, plan: iv.SHOPIFY_PLAN }),

  // Store the plan information in a cookie. This allows the frontend to automatically hit this endpoint after
  // the authn/authz sequence completes as a result of triggering this API in the first place.
  // There is probably a better way to do this.
  // @see https://github.com/Shopify/shopify-app-js/issues/470
  (req, res, next) => {
    res.setHeader('Set-Cookie', [`plan=${req.query.plan}; Path=/;`]);
    next();
  },

  // Instead of using the built-in shopify.validateAuthenticatedSession(), we use our own middleware because
  // the build-in one performs a 302 redirect to another domain when the token is invalid, which causes the
  // XHR request initiated by the frontend to fail due to CORS.
  // @see https://github.com/Shopify/shopify-app-js/issues/470
  (req, res, next) => {
    try {
      (async () => {
        req.logger.debug({ m: 'Running validateAuthenticatedSession' });

        const sessionId = await shopify.api.session.getCurrentId({
          isOnline: shopify.config.useOnlineTokens,
          rawRequest: req,
          rawResponse: res,
        });

        let session: Session | undefined;
        if (sessionId) {
          session = await shopify.config.sessionStorage.loadSession(sessionId);
        }

        const shop =
          shopify.api.utils.sanitizeShop(req.query.shop as string) ||
          session?.shop;
        const authUrl = `${shopify.config.auth.path}?shop=${shop}`;

        if (session && shop && session.shop !== shop) {
          req.logger.debug({
            m: `Found a session for a different shop in the request: ${JSON.stringify(
              {
                currentShop: session.shop,
                requestShop: shop,
              }
            )}`,
          });

          // Responding with JSON instead of doing a 302 redirect is the only main difference compared to the
          // out-of-the-box shopify.validateAuthenticatedSession() middleware
          return res.status(403).json({ authUrl });
        }

        if (session) {
          req.logger.debug({
            m: `Request session found and loaded: ${session.shop}`,
          });

          if (session.isActive(shopify.api.config.scopes)) {
            req.logger.debug({
              m: `Request session exists and is active: ${session.shop}`,
            });

            if (await hasValidAccessToken(shopify.api, session)) {
              req.logger.debug({
                m: `Request session has a valid access token: ${session.shop}`,
              });
              res.locals.shopify = { ...res.locals.shopify, session };
              return next();
            }
          }
        }

        req.logger.info({
          m: `Session was not valid. Redirecting to ${authUrl}: ${shop}`,
        });

        // Responding with JSON instead of doing a 302 redirect is the only main difference compared to the
        // out-of-the-box shopify.validateAuthenticatedSession() middleware
        return res.status(403).json({ authUrl });
      })();
    } catch (err) {
      handleServerError(res, req.logger)(err);
    }
  },

  // With a valid session, make the Billing API request
  async (req, res) => {
    try {
      const { plan } = req.query;

      const confirmationUrl = await shopify.api.billing.request({
        session: req.locals.shopify.session,
        plan,
        returnUrl: `${process.env.APP_HOST}a/landing/shopify/signup/confirm`,
        isTest: process.env.NODE_ENV === 'development',
      });

      return res.json({ confirmationUrl });
    } catch (err) {
      handleServerError(res, req.logger)(err);
    }
  }
);

```

... and here is the frontend code that initiates it (when "Subscribe" is pressed):


```typescript
const pay = async (shop: string, plan: ShopifyBillingConfigPlan) => {
  const { data, error, response } = await commonFetch(`/api/integration/shopify/pay?shop=${shop}&plan=${plan}`);

  if (response.status === 403 && error?.authUrl !== undefined) {
    // Token invalid; initialize auth flow
    window.location.href = error.authUrl;
  } else if (error !== undefined) {
    // Something went wrong; display error
    state.error = error;
  } else if (data?.confirmationUrl !== undefined) {
    // API succeeded; redirect to confirmation url
    window.location.href = data.confirmationUrl;
  }
};
```

In my dream scenario, the first two middlewares in the route handler above would come out-of-the-box with this package (or something similar to solve the same problem).

**Comment by paulomarg** [comment]
Hey, and sorry it took us so long to get to this. I agree, this is essentially broken right now. Thank you for providing so much detail, it made it much easier to solve it.

Thank you for your patience!
